### PR TITLE
OBSDOCS-213: Loki reliability hardening reference docs

### PR DESCRIPTION
**Pull request by abrennan89**
Version(s):
4.12+

Issue:
https://issues.redhat.com/browse/OBSDOCS-213

Link to docs preview:
- https://64845--docspreview.netlify.app/openshift-enterprise/latest/logging/cluster-logging-loki#logging-loki-reliability-hardening_cluster-logging-loki
- https://64845--docspreview.netlify.app/openshift-enterprise/latest/nodes/scheduling/nodes-scheduler-pod-affinity (no content change, moved to snippet only)

QE review:
- [x] QE has approved this change.

### CODE REVIEW

File: modules/logging-loki-reliability-hardening.adoc [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+----
+apiVersion: loki.grafana.com/v1
+kind: LokiStack
+metadata:
+  name: lokistack-dev
+  namespace: openshift-logging
+spec:
+# ...
+  template:
+    ingester: <1>
```
Comment (abrennan89) [inline modules/logging-loki-reliability-hardening.adoc:29]: show examples of different templates

**Review by periklis — CHANGES_REQUESTED** [review]

File: modules/logging-loki-reliability-hardening.adoc [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            topologyKey: kubernetes.io/hostname <7>
+          weight: 100
+# ...
+----
+<1> The component `template` spec. This value can be `compactor`, `distributor`, `gateway`, `indexGateway`, `ingester`, `querier`, `queryFrontend`, or `ruler`, depending on the component being configured. Multiple components are included in one `LokiStack` custom resource (CR).
+<2> The stanza to configure node anti-affinity.
+<3> The stanza to define a preferred rule.
+<4> The key/value pair (label) that must be matched to apply the rule.
+<5> The label for the component being configured. This value matches the field name from the component `template` spec `<1>`.
+<6> The namespace where the Loki components are deployed.
```
Comment (JoaoBraveCoding) [inline modules/logging-loki-reliability-hardening.adoc:46]: Here it's actually the name of the LokiStack CR that  creates the different resources, not the namespace where they are deployed
```suggestion
<6> The name of the LokiStack instance.
```

File: modules/logging-loki-reliability-hardening.adoc [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+[id="logging-loki-reliability-hardening-default-settings_{context}"]
+== Default Loki settings
+
+The Operator sets default, preferred `podAntiAffinity` rules for all Loki components, which includes the `compactor`, `distributor`, `gateway`, `indexGateway`, `ingester`, `querier`, `queryFrontend`, and `ruler` components.
+
+.Example default preferred rule settings for the ingester component
+[source,yaml]
+----
+apiVersion: loki.grafana.com/v1
```
Comment (kabirbhartiRH) [inline modules/logging-loki-reliability-hardening.adoc:21]: I don't think we need to specify the default settings in the LokiStack CR. the podAntiAffinity rules should be applied  by default when a CR is created for all loki components. For e.g.
```
$ oc get statefulset lokistack-dev-ingester -o json | jq -r '.spec.template.spec.affinity'
{
  "nodeAffinity": {
    "requiredDuringSchedulingIgnoredDuringExecution": {
      "nodeSelectorTerms": [
        {
          "matchExpressions": [
            {
              "key": "kubernetes.io/os",
              "operator": "In",
              "values": [
                "linux"
              ]
            }
          ]
        }
      ]
    }
  },
  "podAntiAffinity": {
    "preferredDuringSchedulingIgnoredDuringExecution": [
      {
        "podAffinityTerm": {
          "labelSelector": {
            "matchLabels": {
              "app.kubernetes.io/component": "ingester",
              "app.kubernetes.io/instance": "lokistack-dev"
            }
          },
          "topologyKey": "kubernetes.io/hostname"
        },
        "weight": 100
      }
    ]
  }
}
```

File: modules/logging-loki-reliability-hardening.adoc [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+kind: LokiStack
+metadata:
+  name: lokistack-dev
```
Comment (kabirbhartiRH) [inline modules/logging-loki-reliability-hardening.adoc:24]: Also, the docs currently about loki have the lokistack name as 'logging-loki' :)

File: modules/logging-loki-reliability-hardening.adoc (hunk 3 above)
Comment (kabirbhartiRH) [reply modules/logging-loki-reliability-hardening.adoc:21]: Here the CR created had below spec 
```
spec:
  managementState: Managed
  rules:
    enabled: true
    namespaceSelector:
      matchLabels:
        openshift.io/cluster-monitoring: 'true'
    selector:
      matchLabels:
        openshift.io/cluster-monitoring: 'true'
  size: 1x.demo
  storage:
    schemas:
      - effectiveDate: '2022-06-01'
        version: v12
    secret:
      name: test
      type: s3
  storageClassName: gp2-csi
  tenants:
    mode: openshift-logging
```

File: modules/logging-loki-reliability-hardening.adoc (hunk 2 above)
Comment (abrennan89) [reply modules/logging-loki-reliability-hardening.adoc]: Ended up removing this default example, since QE pointed out it wasn't clear that this didn't need to be configured

File: modules/logging-loki-reliability-hardening.adoc (hunk 3 above)
Comment (abrennan89) [reply modules/logging-loki-reliability-hardening.adoc]: Removed information about the default config, kept only information about how to override the default config

File: modules/logging-loki-reliability-hardening.adoc (hunk 4 above)
Comment (abrennan89) [reply modules/logging-loki-reliability-hardening.adoc]: Updated, thanks

**Review by periklis — COMMENTED** [review]
LGTM

File: modules/logging-loki-reliability-hardening.adoc [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+:_content-type: REFERENCE
+[id="logging-loki-reliability-hardening_{context}"]
+= Configuring Loki to tolerate node failure
+
+In logging 5.8 and newer versions, the Loki Operator supports setting pod anti-affinity rules to request that pods of the same component are scheduled on different available nodes in the cluster.
+
+include::snippets/about-pod-affinity.adoc[]
+
+The Operator sets default, preferred `podAntiAffinity` rules for all Loki components, which includes the `compactor`, `distributor`, `gateway`, `indexGateway`, `ingester`, `querier`, `queryFrontend`, and `ruler` components.
```
Comment (adellape) [inline modules/logging-loki-reliability-hardening.adoc:13]: Possible candidate for an unordered list, but I also don't know if I actually like that better or not (taking up extra real estate).

File: modules/logging-loki-reliability-hardening.adoc [hunk 6]
```diff
@@ -0,0 +1,39 @@
+// Module included in the following assemblies:
+//
+// * logging/cluster-logging-loki.adoc
+
+:_content-type: REFERENCE
```
Comment (adellape) [inline modules/logging-loki-reliability-hardening.adoc:5]: This module seems a little more `CONCEPT` than `REFERENCE` possibly? Although the "Configuring..." title and the last bit is more procedure-y sounding, too, so I'm not sure.

Comment (abrennan89) [reply modules/logging-loki-reliability-hardening.adoc:5]: ACK, I'd say it's actually more conceptual. I was thinking reference just because we give that config example in there, but really it's just explaining the concept of podaffinity for loki and that you can configure it, rather than giving any real step by step / sample code for configuring it. It would probably be another procedure module if we wanted to explain how users configure specific rules, but that's outside the scope.
TL;DR, I'll update this to a `CONCEPT` 🙂

File: modules/logging-loki-reliability-hardening.adoc [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+[id="logging-loki-reliability-hardening_{context}"]
+= Configuring Loki to tolerate node failure
+
+In logging 5.8 and newer versions, the Loki Operator supports setting pod anti-affinity rules to request that pods of the same component are scheduled on different available nodes in the cluster.
```
[the remark this answers is not in the document]
Comment (abrennan89) [reply modules/logging-loki-reliability-hardening.adoc:9]: ACK, I will add a Jira to follow up on making sure this is standard (I didn't realize it was "later" until very recently 😅).
RE the naming, Logging isn't an official product name AFAIK. We actually have a separate Jira to figure that out, but I think we have an attribute. It's not really correct, and very wordy ("the logging subsystem for OpenShift",  I think 😬) so I was trying to avoid using it since we'll need to replace / reword around it later, but for now I think it's the best way forward.

File: modules/logging-loki-reliability-hardening.adoc (hunk 5 above)
Comment (abrennan89) [reply modules/logging-loki-reliability-hardening.adoc:13]: I think I prefer this one as is 🙂

File: modules/logging-loki-reliability-hardening.adoc [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+      # ...
+        requiredDuringSchedulingIgnoredDuringExecution: <1>
+        - labelSelector:
+            matchLabels: <2>
+              app.kubernetes.io/component: ingester
+          topologyKey: kubernetes.io/hostname
+# ...
+----
+<1> The stanza to define a required rule.
+<2> The key/value pair (label) that must be matched to apply the rule.
```
[the remark this answers is not in the document]
Comment (abrennan89) [reply modules/logging-loki-reliability-hardening.adoc:39]: Great information, thank you!

**Review by adellape — APPROVED** [review]

**Comment by abrennan89** [comment]
I will manually CP this back to 4.11 so that the snippet changes are the same in all active branches, no reason not to do that 🙂
